src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -1179,9 +1179,24 @@ private void getMethodSig(CORINFO_METHOD_STRUCT_* ftn, CORINFO_SIG_INFO* sig, CO
             Get_CORINFO_SIG_INFO(method, sig: sig, scope: null);
         }
 
-        private bool getMethodInfo(CORINFO_METHOD_STRUCT_* ftn, CORINFO_METHOD_INFO* info)
+        private bool getMethodInfo(CORINFO_METHOD_STRUCT_* ftn, CORINFO_METHOD_INFO* info, CORINFO_CONTEXT_STRUCT* context)
         {
             MethodDesc method = HandleToObject(ftn);
+
+            if (context != null && method.IsSharedByGenericInstantiations)
+            {
+                TypeSystemEntity ctx = entityFromContext(context);
+                if (ctx is MethodDesc methodFromCtx)
+                {
+                    method = methodFromCtx;
```

Comment:
```suggestion
                    Debug.Assert(method.GetTypicalMethodDefinition() == methodFromCtx.GetTypicalMethodDefinition());
                    method = methodFromCtx;
```

Reply:
`entityFromContext` may return `MethodBeingCompiled` as a method context - I guess that's not what we want here when we invoke `getMethodInfo` for a random call inside current method - so I guarded the check with `context != contextFromMethodBeingCompiled()`

Reply:
Although, the current CoreCLR is a bit more conservative - it ignores `context == currentMethodContext` and all methoddesc-like generics completely:

![image](https://github.com/dotnet/runtime/assets/523221/823942e1-130e-475e-a6a8-42fb7fe832a3)

so maybe I should align this here as well?

Reply:
If JIT/EE interface method takes a `context` argument, I would expect it to be respected for all cases. People like to write complicated generics. Why not optimize all of them that we can optimize here?

Reply:
FWIW, we rely heavily on generics for optimization in the JIT.
For example:

```
private static delegate*<ref SortingMatch<TInner>, Span<long>, int> 
    SortBy<TEntryComparer, TLookupKey, TFwdIt,TBackIt>(OrderMetadata metadata)
    where TEntryComparer : struct, IEntryComparer, IComparer<UnmanagedSpan>
    where TLookupKey : struct, ILookupKey
    where TFwdIt : struct,  ILookupIterator
    where TBackIt : struct, ILookupIterator
``

And many others where we are trying to get inlining for specialized impl

Reply:
> If JIT/EE interface method takes a `context` argument, I would expect it to be respected for all cases. People like to write complicated generics. Why not optimize all of them that we can optimize here?

Added back to both - checked on CoreCLR and its impact on diffs was quite small compared to class ctx